pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -0,0 +1,173 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"errors"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Node is a set of constraints, compatible pods, and possible instance types that could fulfill these constraints. This
+// will be turned into one or more actual node instances within the cluster after bin packing.
+type Node struct {
+	Constraints         *v1alpha5.Constraints
+	InstanceTypeOptions []cloudprovider.InstanceType
+	Pods                []*v1.Pod
+}
+
+func NewNode(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods ...*v1.Pod) *Node {
+	n := &Node{Constraints: constraints.DeepCopy()}
+
+	provIts := constraints.Requirements.InstanceTypes()
+	for _, it := range instanceTypes {
+		// provisioner must list the instance type by name (defaults to all instance types)
+		if provIts.Has(it.Name()) {
```

Comment:
Consider a shortcircuitting idiom to avoid deeply nesting lines 42-52

```suggestion
		if !provIts.Has(it.Name()) {
		  continue
		}
```